File: scripts/posttool_handler.py
#!/usr/bin/env python3
"""PostToolUse safety net for data-protector plugin.

Two layers of detection:
1. Column name detection — check if protected column names appear as headers
2. Tabular data leak detection — check if output looks like unfiltered multi-column data
   (many rows of delimiter-separated values without [PROTECTED] markers)
"""
import json
import os
import re
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from config_manager import load_config

MASK = "[PROTECTED]"
TABULAR_ROW = re.compile(r"^.+[,\t|].+[,\t|].+$")


def find_leaked_columns(text, config):
    columns = config.get("columns", [])
    mode = config.get("mode", "block")
    case_sensitive = config.get("case_sensitive", False)

    if mode != "block":
        return []

    found = []
    for col in columns:
        pattern = re.escape(col)
        flags = 0 if case_sensitive else re.IGNORECASE
        if re.search(r"\b" + pattern + r"\b", text, flags):
            found.append(col)
    return found


def detect_unfiltered_tabular(text, config):
    """Detect if output contains tabular data that should have been filtered but wasn't.

    Heuristic: if we see 3+ rows of delimiter-separated data AND none contain [PROTECTED],
    this is likely an unfiltered data leak.
    """
    columns = config.get("columns", [])
    if not columns:
        return False

    lines = text.strip().split("\n")
    tabular_lines = [l for l in lines if TABULAR_ROW.match(l.strip())]

    if len(tabular_lines) < 3:
        return False

    has_protected = any(MASK in l for l in tabular_lines)
    if has_protected:
        return False

    case_sensitive = config.get("case_sensitive", False)
    first_tabular = tabular_lines[0]
    flags = 0 if case_sensitive else re.IGNORECASE
    col_matches = 0
    for col in columns:
        if re.search(r"\b" + re.escape(col) + r"\b", first_tabular, flags):
            col_matches += 1

    return col_matches >= 1


def main():
    try:
        input_data = json.load(sys.stdin)
    except json.JSONDecodeError:
        sys.exit(0)

    config = load_config()
    if not config.get("columns"):
        sys.exit(0)

    tool_result = input_data.get("tool_result", "")
    if isinstance(tool_result, dict):
        tool_result = json.dumps(tool_result)

    if not tool_result:
        sys.exit(0)

    text = str(tool_result)
    leaked = find_leaked_columns(text, config)
    tabular_leak = detect_unfiltered_tabular(text, config)

    warnings = []

    if tabular_leak:
        warnings.append(
            "CRITICAL: The output above contains UNFILTERED tabular data with protected columns. "
            "This data bypassed the filter. You MUST NOT display, quote, or reference ANY values "
            "from the protected columns. Discard this output entirely and use the filter command:\n"
            f'python3 "<plugin>/scripts/filter.py" "<file_path>"'
        )
    elif leaked:
        cols_str = ", ".join(leaked)
        warnings.append(
            f"The output above contains protected columns: {cols_str}. "
            f"Do NOT reference, quote, display, or use the actual values from these columns. "
            f"Treat them as confidential."
        )

    if warnings:
        result = {
            "systemMessage": "DATA PROTECTOR WARNING: " + " ".join(warnings)
        }
        print(json.dumps(result))

    sys.exit(0)


if __name__ == "__main__":
    main()

File: scripts/posttool_handler.js
#!/usr/bin/env node
const path = require("path");
const { loadConfig } = require("./config_manager");

function main() {
  let input = "";
  process.stdin.setEncoding("utf-8");
  process.stdin.on("data", (chunk) => (input += chunk));
  process.stdin.on("end", () => {
    let data;
    try {
      data = JSON.parse(input);
    } catch {
      process.exit(0);
    }

    const toolName = data.tool_name || "";
    const toolInput = data.tool_input || {};
    const toolOutput = (data.tool_result || data.tool_output || "").toString();

    if (!toolOutput || toolOutput.length < 10) process.exit(0);

    const config = loadConfig();
    const columns = config.columns || [];
    if (!columns.length && config.mode === "block") process.exit(0);

    const leaked = findLeakedColumns(toolOutput, config);
    const tabularLeak = detectUnfilteredTabular(toolOutput, config);

    if (!leaked.length && !tabularLeak) process.exit(0);

    const warnings = [];
    if (tabularLeak) {
      warnings.push(
        "CRITICAL: The output above contains UNFILTERED tabular data with protected columns. " +
          "This data bypassed the filter. You MUST NOT display, quote, or reference ANY values " +
          'from the protected columns. Discard this output entirely and use the filter command.'
      );
    } else if (leaked.length) {
      warnings.push(
        `The output above contains protected columns: ${leaked.join(", ")}. ` +
          "Do NOT reference, quote, display, or use the actual values from these columns. " +
          "Treat them as confidential."
      );
    }

    if (warnings.length) {
      console.log(
        JSON.stringify({
          systemMessage: "DATA PROTECTOR WARNING:\n" + warnings.join("\n"),
        })
      );
    }
    process.exit(0);
  });
}

function findLeakedColumns(output, config) {
  const columns = config.columns || [];
  const mode = config.mode || "block";
  const caseSensitive = config.case_sensitive || false;

  let checkCols;
  if (mode === "block") {
    checkCols = columns;
  } else {
    return [];
  }

  const leaked = [];
  for (const col of checkCols) {
    const target = caseSensitive ? col : col.toLowerCase();
    const text = caseSensitive ? output : output.toLowerCase();
    if (text.includes(target)) leaked.push(col);
  }
  return leaked;
}

function detectUnfilteredTabular(output, config) {
  const lines = output.split("\n").filter((l) => l.trim());
  if (lines.length < 3) return false;

  const delimiters = ["\t", ",", "|"];
  for (const d of delimiters) {
    const counts = lines.slice(0, 5).map((l) => l.split(d).length);
    if (counts[0] >= 2 && counts.every((c) => c === counts[0])) {
      const hasProtected = !output.includes("[PROTECTED]");
      if (hasProtected) {
        const headerLine = lines[0].toLowerCase();
        const columns = config.columns || [];
        for (const col of columns) {
          if (headerLine.includes(col.toLowerCase())) return true;
        }
      }
    }
  }
  return false;
}

main();
